routing/router.go

```diff
@@ -1938,34 +1941,27 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 	}
 }
 
-// pruneEdgeFailure will attempts to prune an edge from the current available
-// edges of the target payment session in response to an encountered routing
-// error.
-func pruneEdgeFailure(paySession *paymentSession, route *Route,
-	errSource *btcec.PublicKey) {
+func getFailedChannelID(route *Route, errSource *btcec.PublicKey) (
+	uint64, error) {
 
 	// As this error indicates that the target channel was unable to carry
 	// this HTLC (for w/e reason), we'll query the index to find the
 	// _outgoing_ channel the source of the error was meant to pass the
 	// HTLC along to.
-	badChan, ok := route.nextHopChannel(errSource)
-	if !ok {
-		// If we weren't able to find the hop *after* this node, then
-		// we'll attempt to disable the previous channel.
-		prevChan, ok := route.prevHopChannel(
-			errSource,
-		)
-
-		if !ok {
-			return
-		}
+	if badChan, ok := route.nextHopChannel(errSource); ok {
+		return badChan.ChannelID, nil
+	}
 
-		badChan = prevChan
+	// If we weren't able to find the hop *after* this node, then
+	// we'll attempt to disable the previous channel.
+	//
+	// TODO(joostjager): errSource must be the final hop then?
```

Comment:
I think the final node should never report errors that have to do with the next channel/hop, as it doesn't exist. The logic here to prune the channel to the final node could make sense in that light, as the final hop is sending an error that doesn't make sense. Or even prune the final node completely?

Reply:
It's not that they prune error with the next channel/hop, but that they send an error back for the final hop itself. In this case, we need to check the prior channel. Not a blocker, but I think this can be slightly re-written to check if the error source is the final destination, if so then we can return the final hop immediately.

Reply:
What I mean is this: A final hop can't return any failure. For example `FailUnknownNextPeer` cannot come from the final hop. In our current handling of failures, we don't need that failed channel id for any of the 'final hop' failures. If a final hop does return such an error (like `FailUnknownNextPeer`), isn't that reason for pruning the node?

Reply:
Of course it can, for example: `FailUnknownPaymenthash`

Reply:
Sorry, bad English with the word any. Not all failures are expected from a final hop. `FailUnknownPaymenthash` sure, but not `FailUnknownNextPeer`